File: php-hosting/src/counterRegistry.php
<?php

/**
 * Structural validation and canonicalization for counter cells. Mirrors
 * servers/ktor/.../counters/CounterRegistry.kt exactly, including the canonicalization
 * algorithm -- the two must hash identical dims to the identical value, or the same logical
 * cell silently splits into two rows across backends. Keep both in sync; see
 * counterRegistryTest.php for the shared vectors.
 *
 * A counter is (metric, dims) -> n. This registry is what stands between arbitrary SDK input
 * and the `counters` table: an unknown metric, an undeclared dimension, or a value outside the
 * allowed charset is rejected here rather than persisted.
 */

/** v1 metrics: each requires exactly this set of dimension keys -- no more, no fewer. */
function counterMetricRegistry(): array {
    return [
        'screen_transition' => ['from', 'to'],
        'screen_dwell'       => ['screen', 'bucket'],
        'funnel_step'        => ['f', 'rev', 'step'],
        'retention'          => ['cohort', 'day'],
        'session'            => ['bucket'],
        'event'              => ['name'],
        'value'              => ['name'],
        'activation'         => ['cohort'],
        'search'             => ['screen'],
        'search_zero_result' => ['screen'],
        'friction'           => ['screen', 'kind'],
    ];
}

const COUNTER_KEY_REGEX = '/^[a-z0-9_]{1,32}$/';
const COUNTER_VALUE_REGEX = '/^[A-Za-z0-9_.:\/-]{1,128}$/';

/**
 * Validates $dims against the declared shape of $metric and, if valid, canonicalizes and
 * hashes them. Returns ['valid' => true, 'canonical' => string, 'hash' => string] or
 * ['valid' => false, 'reason' => string].
 */
function counterValidate(string $metric, array $dims): array {
    $registry = counterMetricRegistry();
    if (!array_key_exists($metric, $registry)) {
        return ['valid' => false, 'reason' => 'unknown_metric'];
    }
    $required = $registry[$metric];
    $keys = array_keys($dims);
    sort($required);
    $sortedKeys = $keys;
    sort($sortedKeys);
    if ($sortedKeys !== $required) {
        return ['valid' => false, 'reason' => 'dims_mismatch'];
    }
    foreach ($dims as $key => $value) {
        if (!is_string($key) || !preg_match(COUNTER_KEY_REGEX, $key)) {
            return ['valid' => false, 'reason' => 'dim_key_invalid'];
        }
        if (!is_string($value) || !preg_match(COUNTER_VALUE_REGEX, $value)) {
            return ['valid' => false, 'reason' => 'dim_value_invalid'];
        }
    }
    $canonical = counterCanonicalize($dims);
    return ['valid' => true, 'canonical' => $canonical, 'hash' => hash('sha256', $canonical)];
}

/**
 * Sorted-key join with ASCII unit/record separators (\x1F, \x1E) between value/key and
 * between pairs. Values are charset-restricted above so neither separator can ever appear
 * inside a key or value, making this join unambiguous without percent-encoding.
 */
function counterCanonicalize(array $dims): string {
    ksort($dims);
    $parts = [];
    foreach ($dims as $key => $value) {
        $parts[] = $key . "\x1F" . $value;
    }
    return implode("\x1E", $parts);
}

File: servers/ktor/src/main/kotlin/com/quietmetrix/server/counters/CounterRegistry.kt
package com.quietmetrix.server.counters

import java.security.MessageDigest

/**
 * Structural validation and canonicalization for counter cells. A counter is
 * `(metric, dims) -> n`; this registry is what stands between arbitrary SDK input and the
 * `counters` table, so it is the enforcement point for "no free-form strings except in
 * threshold-gated fields": an unknown metric, an undeclared dimension, or a value outside the
 * allowed charset is rejected here rather than persisted.
 *
 * Mirrors php-hosting/src/counterRegistry.php exactly, including the canonicalization
 * algorithm — the two must hash identical dims to the identical value, or the same logical
 * cell silently splits into two rows across backends. Keep both in sync; see
 * CounterRegistryTest.kt for the shared vectors.
 */
object CounterRegistry {

    /** One declared metric: its exact required set of dimension keys. */
    data class MetricSpec(val name: String, val requiredDims: Set<String>)

    /**
     * v1 metrics. Each requires exactly this set of dims — no more, no fewer — so canonical
     * form never needs to encode which keys were present, only their values in sorted-key
     * order.
     */
    val metrics: Map<String, MetricSpec> = listOf(
        MetricSpec("screen_transition", setOf("from", "to")),
        MetricSpec("screen_dwell", setOf("screen", "bucket")),
        MetricSpec("funnel_step", setOf("f", "rev", "step")),
        MetricSpec("retention", setOf("cohort", "day")),
        MetricSpec("session", setOf("bucket")),
        MetricSpec("event", setOf("name")),
        MetricSpec("value", setOf("name")),
        MetricSpec("activation", setOf("cohort")),
        MetricSpec("search", setOf("screen")),
        MetricSpec("search_zero_result", setOf("screen")),
        MetricSpec("friction", setOf("screen", "kind")),
    ).associateBy { it.name }

    private val KEY_REGEX = Regex("^[a-z0-9_]{1,32}$")
    private val VALUE_REGEX = Regex("^[A-Za-z0-9_.:/-]{1,128}$")

    /** ASCII unit/record separators. Values are charset-restricted above so neither can ever
     *  appear inside a key or value, making this join unambiguous without percent-encoding. */
    private const val UNIT_SEP = '\u001F'
    private const val RECORD_SEP = '\u001E'

    sealed class ValidationResult {
        data class Valid(val canonicalDims: String, val dimsHash: String) : ValidationResult()
        data class Invalid(val reason: String) : ValidationResult()
    }

    /** Validates [dims] against the declared shape of [metric] and, if valid, canonicalizes
     *  and hashes them. The hash is the grouping key callers use to upsert a counter cell. */
    fun validate(metric: String, dims: Map<String, String>): ValidationResult {
        val spec = metrics[metric] ?: return ValidationResult.Invalid("unknown_metric")
        if (dims.keys != spec.requiredDims) return ValidationResult.Invalid("dims_mismatch")
        for ((key, value) in dims) {
            if (!KEY_REGEX.matches(key)) return ValidationResult.Invalid("dim_key_invalid")
            if (!VALUE_REGEX.matches(value)) return ValidationResult.Invalid("dim_value_invalid")
        }
        val canonical = canonicalize(dims)
        return ValidationResult.Valid(canonical, sha256Hex(canonical))
    }

    private fun canonicalize(dims: Map<String, String>): String =
        dims.entries
            .sortedBy { it.key }
            .joinToString(RECORD_SEP.toString()) { (key, value) -> "$key$UNIT_SEP$value" }

    private fun sha256Hex(input: String): String {
        val digest = MessageDigest.getInstance("SHA-256").digest(input.toByteArray(Charsets.UTF_8))
        return digest.joinToString("") { "%02x".format(it) }
    }
}
